Reduce unsafeness in assorted HTML elements

annevk opened the pull request:
#### a8cca3e460dc7e9e1a9894c05fefe3b62bbc3284
<pre>
Reduce unsafeness in assorted HTML elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=301342">https://bugs.webkit.org/show_bug.cgi?id=301342</a>

Reviewed by Ryosuke Niwa and Sam Weinig.

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines</a> to a
couple more HTML elements.

Canonical link: <a href="https://commits.webkit.org/302165@main">https://commits.webkit.org/302165@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e100ed8e2125f6989e70e2f46c399a20b9bc27a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128191 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/475 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39023 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135589 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79679 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/52dbb4be-b5a6-493c-8180-bf3ae038ff6f) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130063 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/400 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/346 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97581 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65479 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/86ceec30-0b53-4cf6-ae7f-97084112c6ff) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131139 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/253 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114837 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78152 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b99dcc6e-1fb4-423a-b376-180ede9979e4) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/243 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32945 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78863 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108621 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33429 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138042 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/326 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/307 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106111 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/358 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111179 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105894 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26992 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/252 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29733 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52576 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/372 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/62617 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/279 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/352 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/337 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/html/HTMLFieldSetElement.cpp:
```diff
@@ -87,11 +87,12 @@ static void updateFromControlElementsAncestorDisabledStateUnder(HTMLElement& sta
 
     while (it != range.end()) {
-        if (auto* listedElement = it->asValidatedFormListedElement())
+        RefPtr element = *it;
```
Should this be `Ref element`? Given `element` is not null checked in the next line, it seems like it.

weinig commented on Source/WebCore/html/HTMLFormElement.cpp:
```diff
@@ -159,7 +159,7 @@ unsigned HTMLFormElement::length() const
         RefPtr element { weakElement.get() };
         ASSERT(element);
-        auto* listedElement = element->asFormListedElement();
+        RefPtr listedElement = element->asFormListedElement();
```
This should probably a Ref<> as well (given the assert on the next line).

annevk replied:
This returns a pointer. We could dereference it earlier and remove the ASSERT, but I'm not sure we should make that change as part of this?

weinig commented on Source/WebCore/html/HTMLFormElement.cpp:
```diff
@@ -570,12 +570,12 @@ unsigned HTMLFormElement::formElementIndex(FormListedElement& listedElement)
-    for (auto& element : descendants) {
-        if (&element == &listedHTMLElement)
+    for (Ref element : descendants) {
+        if (element.ptr() == listedHTMLElement.ptr())
```
If both of these are Ref<> now, you probably don't need the `.ptr()` for either one.

weinig commented on Source/WebCore/html/HTMLFormElement.cpp:
```diff
@@ -970,8 +970,8 @@ RefPtr<DOMFormData> HTMLFormElement::constructEntryList(RefPtr<HTMLFormControlEl
         submitter->setActivatedSubmit(true);
     
     for (auto& control : this->copyListedElementsVector()) {
```
A little surprised this doesn't want to be `for (Ref control : ...)`.

annevk replied:
I undid the other change where I did that as I didn't realize that `auto&` is different from `Ref`. It's `Ref&` and we probably don't want to change that.

weinig commented on Source/WebCore/html/HTMLTableElement.cpp:
```diff
@@ -547,7 +548,7 @@ const MutableStyleProperties* HTMLTableElement::additionalCellStyle() const
 }
 
-static MutableStyleProperties* leakGroupBorderStyle(int rows)
+static MutableStyleProperties* leakGroupBorderStyle(bool rows)
```
Whoa.

annevk committed:
Reduce unsafeness in assorted HTML elements
https://bugs.webkit.org/show_bug.cgi?id=301342

Reviewed by Ryosuke Niwa and Sam Weinig.

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines to a
couple more HTML elements.

Canonical link: https://commits.webkit.org/302165@main
